### DISCUSSION TITLE
Import data into table from .CSV string (in memory)

### DISCUSSION THREAD

**Discussion by pkese** [duckdb/duckdb#3998] [General]
Hi,

I'm getting some .csv formatted data over network connection and would like to import it to a DuckDB table.

`read_csv_auto` or `COPY` require me to store that data onto disk and read it back as a file.

Is there a way to do that without the extra step of storing that csv data to a file (e.g. pass a memory buffer containing .csv bytes directly to the API)?

**Comment by Mytherin** [comment]
`read_csv_auto` can be used directly in a query, no? Wouldn't `SELECT * FROM read_csv_auto('http://...')` work?

You could also create a temporary table, which will materialize the result only in memory rather than on disk.

**Reply by pkese** [reply]
But *I am* trying to create a temporary table.  
My problem is how to fill that table from csv data that resides in a memory buffer.

`http://` unfortunately does not help, because I retrieve that csv data from some other service which doesn't speek http.

**Reply by Mytherin** [reply]
Which API are you using? From C++ you could use the Appender to do so.

**Reply by Mytherin** [reply]
Ah, nevermind, I understand your question now. There is no direct support for reading CSV files from memory currently.

**Reply by pkese** [reply]
I assume that would probably require me to parse the csv and do the type marshalling, constructing dates, etc.

Would it be a possible workaround to stream data through a named pipe?
Or create a temporary memory mapped file?

**Reply by pkese** [reply]
Hint:  
Clickhouse has this nice syntax, where you can specify `FORMAT` after `INSERT INTO` and then just concatenate data in that format after the query.
And it applies to all formats that the database understands.

```sql
INSERT INTO t (id,name) FORMAT TabSeparated
11  Hello, world!
22  Qwerty
```

**Comment by NickCrews** [comment]
Chiming in here that I have a similar need. I am trying to restore a postgres dump file (custom format) straight into duckdb, avoiding going through a postgres instance. The dump file effectively stores the data as a compressed TSV, so the raw format I am receiving is a TSV from the streaming gzip decompressor. I am doing this in rust.

I am decompressing ~100gb of data, so if I understand correctly the the clickhouse API of
```
INSERT INTO t (id,name) FORMAT TabSeparated
11  Hello, world!
22  Qwerty
```
won't work because it would require me to materialize the entire 100gb of data in memory as a sql string before handing it off to duckdb.

I think the api I would want would be like "mocking the file system". Here is what [pglite does](https://pglite.dev/docs/api#dev-blob):

```javascript
await pg.query("COPY my_table FROM '/dev/blob';", [], {
  blob: MyBlob,
})
```

I would imagine something similar, except in addition to a fully materialized MyBlob, we should also support something that implements a `read()`-like API. Also should do this for writing a byte stream as well.

I think I would prefer to decouple the fs registration from the execution:

```typescript
const src = {
     // emulating the rust Read trait here
    read(buffer: number[]): number
}
const sink = {
     // emulating the rust Write trait here
    write(buffer: number[]): number
}
conn.register_file("/dev/src", src)
conn.register_file("/any/path/sink", sink)
conn.query("COPY (select a, b, c from read_csv('/dev/src')) TO '/any/path/sink'")
```
